### DISCUSSION TITLE
[Chapter 5] 팩토리 메서드

### DISCUSSION THREAD

**Discussion by firewoody237** [cowaine/is-my-code-that-strange#25] [Chapter 04-05 (불변, 응집도)] [answered]
코드 5.7의 예시에서 아래와 같은 팩토리 메서드를 사용했는데요,
```java
static GiftPoint forStandardMembership() {
  return new GiftPoint(STANDARD_MEMBERSHIP_POINT);
}
```

`GiftPoint`는 코드상 매우 가벼운 도메인 코드라서 가능하다고 생각합니다.
다만 실제 사용되는 코드는 `value`만이 아닌 여러 인스턴스변수를 탐욕스럽게 가지고 있는게 대부분이더라구요.
그렇게 되면 보통 다양한 생성자들이 추가되기 마련인데..

문득 파라미터의 경우(예시에서의 `XX_MEMBERSHIP_POINT`)는 `ENUM` 등으로 충분히 제한할 수 있을 것 같은데, 생성자의 장점을 버릴 필요가 있을까요?

**Comment by 0sunset0** [comment, marked as answer]
예시 코드에서 생성자가 아닌 팩토리 메서드를 사용하여 얻은 장점은 
1. `직관적인 이름`을 가져 코드 사용자 입장에서 실수 없이 객체를 생성할 수 있고
2. `STANDARD_MEMBERSHIP_POINT`와 `PREMIUM_MEMBERSHIP_POINT`의 값을 외부로부터 숨길 수 있다는 것 같습니다.

만약 외부에서 `new GiftPoint(STANDARD_MEMBERSHIP_POINT)`와 같이 생성자로 객체를 생성하면 
비즈니스 상수인 `STANDARD_MEMBERSHIP_POINT`를 외부가 알고 있게 됩니다. 
enum으로 만들어서 enum 값을 받아도 import를 받으니 의존한다고 할 수 있을 것 같아요.

비즈니스 요구 사항에 맞게 적절하게 팩토리 메서드를 만들면 되지 않을까 싶습니다. 만약 메서드의 개수가 너무 많아진다면 클래스의 역할 분리를 고려해볼 수 있을 것 같아요.


[그 외 팩토리 메서드 장점]
2. 생성자와 달리 팩토리 메서드는 하위 타입을 반환할 수 있습니다. 하위 타입을 반환하고 상위 타입으로 바꿔 사용한다면 구현을 숨길 수 있습니다.
3. new 키워드는 항상 새로운 객체를 생성하지만, 팩토리 메서드는 항상 새로운 객체를 반환하지 않아도 되고 
객체 생성 방법을 다양하게 조작할 수 있습니다! 자바의 Calendar 클래스에서는 어댑터 패턴을 사용하여 객체를 생성하고 있네요

**Reply by firewoody237** [reply]
아 위의 비즈니스 상수가 외부에 노출안되고/Enum으로 노출되고는 차이가 크겠네요!
생성자를 통해 구현을 노출시킬지 / 팩토리 메서드로 구현을 숨길지 Trade-off 겠군요. 감사합니다!
